### PR TITLE
K8SPXC-599 Add recovery to specific GTID

### PR DESCRIPTION
**Pull request by nonemax**
[K8SPXC-599](https://jira.percona.com/browse/K8SPXC-599) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

File: e2e-tests/pitr/run [hunk 1]
```diff
@@ -18,11 +18,13 @@ run_recovery_check_pitr() {
     local compare=$4
     local time_now=$5
     local dest=$6
+    local gtid=$7
```
Comment (delgod) [inline e2e-tests/pitr/run:21]: can you please run `shfmt -bn -ci -s -w e2e-tests/pitr/run` and commit file again?

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| storage | passed |
| big-data | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| limits | passed |
| upgrade-proxysql | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| smart-update | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| pitr | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/801/commits/041bc137ff9fc99a5ac92d357315390b36a03c22
image: `perconalab/percona-xtradb-cluster-operator:PR-801-041bc137`

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/pxc/v1/pxc_prestore_types.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ type PITR struct {
 	Type         string           `json:"type"`
 	Date         string           `json:"date"`
 	GTIDSet      string           `json:"gtidSet"`
+	GTID         string           `json:"gtid"`
```
[the remark this answers is not in the document]
Comment (nonemax) [reply pkg/apis/pxc/v1/pxc_prestore_types.go:30]: yes, approved by @spron-in

Comment (hors) [inline pkg/apis/pxc/v1/pxc_prestore_types.go:30]: if we decided to add new field we need to add it to `deploy/backup/restore.yaml` too

**Review by hors — CHANGES_REQUESTED** [review]

File: pkg/apis/pxc/v1/pxc_prestore_types.go (hunk 2 above)
Comment (nonemax) [reply pkg/apis/pxc/v1/pxc_prestore_types.go:29]: https://github.com/percona/percona-xtradb-cluster-operator/pull/801/commits/1366152419a43bc8f62c50ee14d339469751fafc

File: cmd/pitr/recoverer/recoverer.go [hunk 3]
```diff
@@ -330,9 +334,28 @@ func (r *Recoverer) setBinlogs() error {
...
+				r.gtidSet = set
+			}
+			if len(r.gtidSet) == 0 {
+				continue
+			}
+		}
+
 		if sourceID != strings.Split(binlogGTIDSet, ":")[0] {
 			continue
 		}
```
[the remark this answers is not in the document]
Comment (nonemax) [reply cmd/pitr/recoverer/recoverer.go]: https://github.com/percona/percona-xtradb-cluster-operator/pull/801/commits/1366152419a43bc8f62c50ee14d339469751fafc

File: e2e-tests/pitr/run (hunk 1 above)
Comment (nonemax) [reply e2e-tests/pitr/run]: https://github.com/percona/percona-xtradb-cluster-operator/pull/801/commits/7ff994d26779ab25ce61ab7f00c9439d5fa70bab

File: cmd/pitr/recoverer/recoverer.go [hunk 4]
```diff
@@ -333,6 +335,24 @@ func (r *Recoverer) setBinlogs() error {
...
+			if subResult != binlogGTIDSet {
+				set, err := getExtendGTIDSet(binlogGTIDSet, r.gtid)
+				if err != nil {
+					return errors.Wrap(err, "get gtid set for extend")
+				}
+				r.gtidSet = set
+			}
+			if len(r.gtidSet) == 0 {
+				continue
+			}
```
Comment (delgod) [inline cmd/pitr/recoverer/recoverer.go:353]: why we should *not* search r.startGTID in binlog which has stopGTID?

**Review by delgod — CHANGES_REQUESTED** [review]

File: cmd/pitr/recoverer/recoverer.go (hunk 4 above)
Comment (nonemax) [reply cmd/pitr/recoverer/recoverer.go:353]: here we just skip files witch didn't have stopGTID

**Review by delgod — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Pull request merged**